activation/nipost.go

```diff
@@ -349,7 +378,57 @@ func membersContainChallenge(members []types.Member, challenge types.Hash32) (ui
 	return 0, fmt.Errorf("challenge is not a member of the proof")
 }
 
-func (nb *NIPostBuilder) getBestProof(ctx context.Context, challenge types.Hash32) (types.PoetProofRef, *types.MerkleProof, error) {
+// TODO(mafa): remove after next poet round; https://github.com/spacemeshos/go-spacemesh/issues/4753
+func (nb *NIPostBuilder) addPoet111ForPubEpoch1(ctx context.Context) error {
+	// because poet 111 had a hardware issue when challenges for round 0 were submitted, no node could submit to it
+	// 111 was recovered with the poet 110 DB, so all successful submissions to 110 should be able to be fetched from there as well
+
+	client111, ok := nb.poetProvers["https://poet-111.spacemesh.network"]
+	if !ok {
+		// poet 111 is not in the list, no action necessary
+		return nil
+	}
+
+	nb.log.Info("pub epoch 1 mitigation: poet 111 is in the list of poets, adding it to the state as well")
+	client110 := nb.poetProvers["https://poet-110.spacemesh.network"]
+
+	ID110, err := client110.PoetServiceID(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get poet 110 id: %w", err)
+	}
+	ID111, err := client111.PoetServiceID(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get poet 111 id: %w", err)
+	}
+
+	if slices.IndexFunc(nb.state.PoetRequests, func(r types.PoetRequest) bool { return bytes.Equal(r.PoetServiceID.ServiceID, ID111.ServiceID) }) != -1 {
+		nb.log.Info("poet 111 is already in the state, no action necessary")
+		return nil
+	}
+
+	poet110 := slices.IndexFunc(nb.state.PoetRequests, func(r types.PoetRequest) bool {
+		return bytes.Equal(r.PoetServiceID.ServiceID, ID110.ServiceID)
+	})
+	if poet110 == -1 {
+		return fmt.Errorf("poet 110 is not in the state, cannot add poet 111")
+	}
+	poet111 := nb.state.PoetRequests[poet110]
+	poet111.PoetServiceID.ServiceID = ID111.ServiceID
+	nb.state.PoetRequests = append(nb.state.PoetRequests, poet111)
+	nb.persistState()
+	nb.log.Info("pub epoch 1 mitigation: poet 111 added to the state")
+	return nil
+}
```

Comment:
Why not just query all `nb.poetProvers` instead (not the ones on `nb.state.PoetRequests`)? If we are not in the members of a poet proof, we simply discard it. The change would be much smaller and fewer places for mistakes.

Reply:
I tried the way you describe and the change looked bigger and more involved (i.e. changes in multiple places instead of just one). This way it is literally only the 40 lines you highlighted.

Going the route of not looking at `nb.state.PoetRequests` but hardcode to query all would require changes in multiple places and also runs the risk of all nodes (even ones that didn't submit) to query the PoET when time comes, which will put a much bigger load on the server.

This way the change is contained to this one function, that updates the state. It also only causes nodes to fetch from PoET 111 if they were previously able to submit to PoET 110 since otherwise they wouldn't be contained in the returned proof anyway and would just needlessly cause load on the PoET.